source/utilities.cc

```diff
@@ -95,225 +195,111 @@ namespace aspect
+                                                                          property_name);
 
-              // Ensure that each entry has the correct form.
-              AssertThrow (key_and_value.size() == 2,
-                           ExcMessage ("The format for mapped "
-                                       + property_name
-                                       + "requires that each entry has the "
-                                       "form `<id> : <value>' "
-                                       ", but the entry <"
-                                       + field_entry
-                                       + "> does not appear to follow this pattern."));
-
-              // If there is one entry in the list the keyword "all" must be found.
-              if ((field_entries.size() == 1) && (n_fields != 1))
-                {
-                  AssertThrow (key_and_value[0] == "all",
-                               ExcMessage ("There is only one "
-                                           + property_name
-                                           + " value given. The keyword `all' is "
-                                           "expected but is not found. Please "
-                                           "check your "
-                                           + property_name
-                                           + " list."));
-
-                  const std::vector<std::string> values = dealii::Utilities::split_string_list(key_and_value[1], '|');
-
-                  // Assign all the elements to the "all" value
-                  for (unsigned int field_index=0; field_index<n_fields; ++field_index)
-                    {
-                      for (const auto &value : values)
-                        {
-                          return_map[field_index].push_back(Utilities::string_to_double(value));
-
-                          if (store_structure)
-                            ++(*n_values_per_key)[field_index];
-                        }
+      // Second: Now check that the structure of the map is as expected
+      {
+        const bool check_structure = (n_values_per_key && n_values_per_key->size() != 0);
+        const bool store_structure = (n_values_per_key && n_values_per_key->size() == 0);
+        std::vector<unsigned int> values_per_key(n_fields, 0);
+
+        if (check_structure)
+          AssertThrow(n_values_per_key->size() == n_fields,
+                      ExcMessage("When providing an expected structure for input parameter " + property_name + " you need to provide "
+                                 + "as many entries in the structure vector as there are input field names (+1 if there is a background field). "
+                                 + "The current structure vector has " + std::to_string(n_values_per_key->size()) + " entries, but there are "
+                                 + std::to_string(n_fields) + " field names." ));
+
+        for (const auto &entry: parsed_map)
```

Comment:
Would you be okay with renaming entry to map_entry? I find it hard to remember what entry.first is, so maybe that would help.